clcache.py

```diff
@@ -1528,25 +1459,38 @@ def processCompileRequest(cache, compiler, args):
     printTraceStatement("Expanded commandline '{0!s}'".format(cmdLine))
 
     try:
-        sourceFiles, objectFile = CommandLineAnalyzer.analyze(cmdLine)
+        sourceFiles, objectFiles = CommandLineAnalyzer.analyze(cmdLine)
 
-        if len(sourceFiles) > 1:
-            return reinvokePerSourceFile(cmdLine, sourceFiles, environment), '', ''
-        else:
-            assert objectFile is not None
-            if 'CLCACHE_NODIRECT' in os.environ:
-                returnCode, compilerOutput, compilerStderr, cleanupRequired = \
-                    processNoDirect(cache, objectFile, compiler, cmdLine, environment)
-            else:
-                returnCode, compilerOutput, compilerStderr, cleanupRequired = \
-                    processDirect(cache, objectFile, compiler, cmdLine, sourceFiles[0])
-            printTraceStatement("Finished. Exit code {0:d}".format(returnCode))
+        baseCmdLine = []
+        for arg in cmdLine:
+            if arg not in sourceFiles:
+                baseCmdLine.append(arg)
```

Comment:
Testing for membership in lists in lists [is O(n)](https://wiki.python.org/moin/TimeComplexity), so this loop scales with O(n^2) with respect to the number of source files on the command line.

However, a quick experiment suggests that this really only starts to hurt for numbers of source files which may not be realistic (for 1500 source files, the loop takes 77ms here, with 15000 source files it's 1567ms). It may not be worth it, but using a `set` would improve this to O(n). I'm not even sure whether you could plausibly invoke `cl.exe` with a few thousand source files, maybe via response files?